contracts/tips/src/state.rs

```diff
@@ -0,0 +1,311 @@
+use crate::error::ContractError;
+use crate::msg::ServiceFee;
+use crate::utils::sum_coins_sorted;
+use cosmwasm_std::{Addr, Coin, Decimal, Uint128};
+use cw_storage_plus::{Item, Map};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use std::collections::VecDeque;
+use std::convert::TryFrom;
+use std::ops::{Div, Mul};
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum StateServiceFee {
+    Fixed { amount: Vec<Coin> },
+    Percentage { value: Decimal },
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct Config {
+    pub admin: Addr,
+    pub subspace_id: u64,
+    pub service_fee: Option<StateServiceFee>,
+    pub tips_history_size: u32,
+}
+
+pub type SentTip = (Addr, Vec<Coin>, u64);
+pub type ReceivedTip = (Addr, Vec<Coin>, u64);
+pub type PostTip = (Addr, Vec<Coin>);
+
+pub const CONFIG: Item<Config> = Item::new("config");
+pub const SENT_TIPS_HISTORY: Map<Addr, VecDeque<SentTip>> = Map::new("sent_tips_history");
+pub const RECEIVED_TIPS_HISTORY: Map<Addr, VecDeque<ReceivedTip>> =
+    Map::new("received_tips_history");
+pub const POST_TIPS_HISTORY: Map<u64, VecDeque<PostTip>> = Map::new("post_tips_history");
```

Comment:
This solution would store duplicated `Vec<Coin>` in each history list.
Can we have a TIPS table to store everything needed, then others only store the index like `tip_id` so that the history data would take less space?

The structure would be like:
TIPS
| column | data |
|--|--|
| tip_id | u64 |
| data |  (Addr, Vec<Coin>, u64) |

SENT_TIPS_HISTORY
| column | data |
|--|--|
| sender | Addr |
| tip_id | u64 |

RECEIVED_TIPS_HISTORY
| column | data |
|--|--|
| receiptient | Addr |
| tip_id | u64 |

POST_TIPS_HISTORY
| column | data |
|--|--|
| post_id | u64 |
| tip_id | u64 |

What do you think?

Reply:
Yes this is the most space efficent way and the previous implementation was something like what you are describing, the problem is keeping the `TIPS` Map in sync requires a lot of compuation and this means greater execution cost.
Lets assume that Alice sent a tip to Bob, after adding the tips inside the `TIPS`, `SENT_TIPS_HISTORY`, `RECEIVED_TIPS_HISTORY` maps the contract need to check if the sent tips history of Alice is greater then the allowed history size if so in order to safely remove the tip the contract needs to iterate over Alice's history tips that execeed the allowed history size and check that the tip thet needs to be removed is also excedining the allowed size of who received that tip, then the contract need to do the same for Bob.
You can have a look of this behaviour in code [here](https://github.com/desmos-labs/desmos-contracts/blob/09d15f3749e93d6f6bcf9b0ea611db0b54337041/contracts/tips/src/contract.rs#L194). 
BTW while writing this i realized that this can be fixed with a reference counter instead of do the search :joy:, I will change the contract to implement such behaviour. 
Also in the `TIPS` map as key I will use a String that will be `block_heigth` plus an index that counts the amount of tips received in the same block and as value a `Tip` structure like this: 
```rust
struct Tip {
    pub sender: Addr,
    pub receiver: Addr,
    pub post_id: Option<u64>,
    pub block_height: u64,
    pub block_index: u64,
} 
```
While for the histories I will keep what you suggest as key and as value I will use a `VecDequeue` to implement the history